src/p4p/test/system_client_server.py

```diff
@@ -0,0 +1,631 @@
+        self.assertIsInstance(counter, ntfloat)
+        self.assertEqual(counter, 0.0)
+        self.assertEqual(counter.severity, 0)
+        self.assertEqual(counter.raw.getID(), "epics:nt/NTScalar:1.0")
+
+    # Equivalent to:
+    #   pvget TESTCOUNTER.counter -r ""
+    def testGetDottedNTScalarNoUnwrap(self):
+        counter = self.ctxt.get("TESTCOUNTER.counter")
+        self.assertEqual(counter.getID(), "epics:nt/NTScalar:1.0")
+        self.assertEqual(counter.value, 0.0)
+        self.assertEqual(counter.timeStamp.getID(), "time_t")
+        self.assertEqual(counter.alarm.getID(), "alarm_t")
+        self.assertEqual(counter.alarm.severity, 0)
+
+    def assertCounter(self, value):
+        if self.MALCOLM:
+            counter = self.ctxt.get("TESTCOUNTER.counter").value
+        else:
+            counter = self.counter_structure.counter.value
+        self.assertEqual(counter, value)
+
+    # Should be equivalent to (except pvput doesn't do the right thing):
+    #   pvput TESTCOUNTER -r "counter.value" 5
+    def testPut(self):
+        self.assertCounter(0)
+        self.ctxt.put("TESTCOUNTER", 5, "counter.value")
+        self.assertCounter(5)
+        self.ctxt.put("TESTCOUNTER", 0, "counter.value")
+        self.assertCounter(0)
+
+    # Equivalent to:
+    #   pvput TESTCOUNTER.counter 5
+    def testPutDotted(self):
+        self.assertCounter(0)
+        # TODO: pvput defaults the request to "value", should p4p do the same?
+        self.ctxt.put("TESTCOUNTER.counter", 5, "value")
+        self.assertCounter(5)
+        self.ctxt.put("TESTCOUNTER.counter", 0, "value")
+        self.assertCounter(0)
+
+    # Equivalent to:
+    #   pvget -m TESTCOUNTER -r ""
+    def testMonitorEverythingInitial(self):
+        q = Queue()
+        m = self.ctxt.monitor("TESTCOUNTER", q.put)
+        self.addCleanup(m.close)
+        counter = q.get()
+        self.assertStructureWithoutTsEqual(str(counter), str(counter_expected))
+        self.assertTrue(counter.asSet().issuperset({
+            "meta", "meta.fields", "counter", "zero"}))
+        self.ctxt.put("TESTCOUNTER.counter", 5, "value")
+        counter = q.get()
+        self.assertEqual(counter.counter.value, 5)
+        if self.MALCOLM:
+            # bitsets in pvaPy don't work, so it is everything at the moment
+            self.assertTrue(counter.asSet().issuperset({
+                "meta", "meta.fields", "counter", "zero"}))
+        else:
+            # TODO: timeStamp userTag didn't change value, but timeStamp did,
+            # should it be in the bitset?
+            self.assertEqual(counter.asSet(),
+                             {"counter.value", "counter.timeStamp",
+                              "counter.timeStamp.secondsPastEpoch"
+                              "counter.timeStamp.nanoseconds"})
+        self.ctxt.put("TESTCOUNTER.counter", 0, "value")
+        counter = q.get()
+        # TODO: Is this true even if you get a subset of changes?
```

Comment:
I'm not sure I understand the question.  In part because I've only glanced at assertStructureWithoutTsEqual(), which I think is testing structure equality of all fields except for the timeStamp sub-structure.  Without having run it (yet) It looks like you're writing the counter back to zero, and then testing that the associated monitor update shows zero.  So I think my answer is "yes".

Reply:
Good. So monitor updates (Value instances) contain the entire structure, even if the particular update only updated a single field. The changed() and asSet() functions allow you to work out what has changed since the last time you got a value.

Reply:
Hmm, what happens if you monitor a structure with a large unchanging array in it, and some scalar bits that change rapidly? I assume pvAccess doesn't send the array each time, so does each Value that the user is passed contain a reference to the same numpy.ndarray as the initial subscription update?

Reply:
PVFields not marked as changed are not serialized.  So, no extra network traffic.  Accounting for arrays in the c++ api is via. ```class shared_vector<const T>``` which is conceptually similar to ```numpy.ndarray```.  So the only "copy" is a copy of this reference.  Actual ndarrays are created each time the value of a field is actually needed.

So not the same ndarray, but the shared_vector<>s point to the same memory.  The main point being that the array itself is not copied (unless it changes).

https://github.com/mdavidsaver/p4p/blob/262f3a4c8ec96fda35f2f09270fcd550f30f3826/src/p4p_value.cpp#L635-L648

Reply:
An important note.  While arrays are extracted from Value by (const) reference, moving an array into Value currently does make a __copy__ of the array itself.  I'm not sure if this will change.

There are some issues here related to the fact that free'ing an ndarray wrapped in shared_vector can involve locking the GIL and running arbitrary python code.  Not something which should happen at places in the c++ code where other locks may be held.  There is also a potential for uncollectable ref. loops involving both python and c++ ref counting.